Call update_device_chats automatically during configure

Hocuri opened the pull request:
update_device_chats() takes about 2 seconds on a modern device (Android) because the
welcome image file has to be written to the disk as a blob. The problem
was that this was done after the progress bar had vanished and before
anything else happened so that I thought that something had gone wrong
multiple times.

The UIs have to remove update_device_chats(), too..

link2xt commented on src/configure/mod.rs:
```diff
@@ -5,6 +5,8 @@ mod auto_outlook;
 mod read_url;
 mod server_params;
 
+use std::time::Duration;
```
There is a warning about unused import.

link2xt commented on deltachat-ffi/deltachat.h:
```diff
@@ -986,6 +986,9 @@ uint32_t        dc_add_device_msg            (dc_context_t* context, const char*
 
 
 /**
+ * DEPRECATED This is now done automatically while configuring.
```
Currently android code calls it on every creation of chatlist.

If it shouldn't do it, then this function is better removed completely, otherwise everyone will forget to remove calls to it.

I don't see any problems with it, except that imported backups from old versions of DC will not get a welcome message. It is even better this way.

After that, maybe "self-chat-added" check and config value could also be removed, since the function is only called once?

Hocuri replied:
> After that, maybe "self-chat-added" check and config value could also be removed, since the function is only called once?

Problem is, configure can be started multiple times (because it fails the first time) and this fn will be called everytime.

> If it shouldn't do it, then this function is better removed completely, otherwise everyone will forget to remove calls to it.

@hpk42 said that we should not break backwards compability where possible (@hpk42 correct?), that's why I didn't up to now. At least compilation will fail but I don't know how obvious it will be that the function was removed.

What I can do is make the body of this fn empty because it does nothing anyway and make update_device_chats() `pub(crate)`.

Simon-Laux replied:
Desktop currently calls this function on each login.
Anyways I agree with @link2xt this time.
Normally I would suggest marking this function as `@deprecated` and to turn it into a no-op, but it looks like we have some breaking changes* in the next version anyway so I believe it makes sense to already remove it now.
(given that we don't forget to mention it in the changelog)
*some PRs that indroduce breaking changes, too: (https://github.com/deltachat/deltachat-core-rust/pull/1968 https://github.com/deltachat/deltachat-core-rust/pull/1965)

Hocuri replied:
Ok, I'll do that as soon as we decided what to do about bots.

Simon-Laux commented:
Keep in mind that **bots don't want to have those chats**, that's a bigger concern than breaking the api by removing the function.

Hocuri commented:
> Keep in mind that **bots don't want to have those chats**

That might be a problem, you're right.

- We could assume that it's no problem. I mean, bots usually just answer on DC_EVENT_MSGS_CHANGED and DC_EVENT_INCOMING_MSG, don't they?
- We could introduce a special "bot" mode that is set via a config. This would have the advantage that if we want to introduce behavior that might cause problems for bots again in the future, we can just use the same config again.

Hocuri commented:
Ok, I now made sure that update_device_chats() does not send any events, which should fix the test failure and will be good for bots. Someone who already has written bots will have to say whether it's enough.

I also removed the deprecated FFI. This also means that this PR now removes more lines than it adds :tada:!

adbenitez commented:
I was reading and thinking: "Finally they will add an 'is bot' flag in core" but well not today :D

Hocuri commented:
@adbenitez it's not too late to decide to add a bot flag for this PR. You have experience with bots, so: This PR still automatically adds two chats on configure, one of them contains "fresh" msgs (but no events are sent). Do we need a flag to disable?

hpk42 commented:
@adbenitez: Why would a bot bother if there are one or two or three automatic chats
after configure? 

On Wed, Oct 07, 2020 at 04:31 -0700, Hocuri wrote:

> @adbenitez it's not too late to decide to add a bot flag for this PR. You have experience with bots, so: This PR still automatically adds two chats on configure, one of them contains "fresh" msgs (but no events are sent). Do we need a flag to disable?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/deltachat/deltachat-core-rust/pull/1957#issuecomment-704873750

r10s commented on src/configure/mod.rs:
```diff
@@ -338,6 +341,11 @@ async fn configure(ctx: &Context, param: &mut LoginParam) -> Result<()> {
     progress!(ctx, 910);
+
+    // Make sure update_device_chats is done before we set the "configured" config
+    // so that no DC_EVENT_MSGS_CHANGED is sent (this would confuse bots and python tests)
```
how is sending DC_EVENT_MSGS_CHANGED avoided by that?

sidenote: i think, in currently implementations these events are not needed, however, in general, it seems to be more consistent to emit them to make sure a maybe already displayed chatlist is updated.

Hocuri replied:
The idea was to be nicer to bots and tests. I think I'll just add a bot flag and fix the tests in another way.

r10s commented:
> update_device_chats() takes about 2 seconds on a modern device (Android) because the welcome image file has to be written to the disk as a blob

i am really wondering why this can take 2 seconds ... on modern devices.

however, apart from that, i would be fine with calling update_device_chats() implicitly. we lose some flexibility (adding os messages before the one from the core, adding core updates messages), which, however, was not used or can be achieved in other ways.

r10s commented on CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+## Unreleased
+
+- breaking change: update_device_chats() was removed. This is now done automatically during configure.
```
nitpicking: usually, we are referring to the ffi-functions here, that would be dc_update_device_chats(), however, thanks for already adding a hint here, this is very helpful for the one writing the next changelog :)

Hocuri commented:
What about this PR? I think I addressed all concerns. Thing is, it's a blocker because the test for https://github.com/deltachat/deltachat-core-rust/pull/1997 need to wait for this PR and I'd like to make a new PR that disables fetch-existing-msgs if the bot flag is set
